### ISSUE TITLE
Pod is OOMKilled with no obvious reason

### ISSUE THREAD

**Issue by strainovic** [k3s-io/k3s#6513]
Hi all,

Thank you for this great product! We started using it daily, and now, we run into issues with Pod's memory consumption.

This is the test Pod we've created:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: memory-demo1
spec:
  containers:
    - name: mem
      image: polinux/stress
      resources:
        requests:
          memory: "50Mi"
        limits:
          memory: "300Mi"
      command: ["stress"]
      args: ["--vm", "2", "--vm-bytes", "250M", "--vm-hang", "0"]
```

When we deploy it on k3s, we're getting OOMKilled all of the time.

k3s

OS Image: Ubuntu 22.04.1 LTS
Container runtime version: containerd://1.6.8-k3s1
kubelet version: v1.24.6+k3s1

```console
k3s -v
k3s version v1.24.7+k3s1 (7af5b167)
go version go1.18.7
```

```console
uname -a
Linux dev-k8s-worker-01 5.15.0-52-generic #58-Ubuntu SMP Thu Oct 13 08:03:55 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

![image](https://user-images.githubusercontent.com/10864240/202201615-60a33655-57b4-454d-a0bf-b8cf0a001006.png)

When we deploy it on microk8s, it works without any issue at 300Mi memory usage.

microk8s

OS Image: Ubuntu 22.04.1 LTS
Container runtime version: containerd://1.6.6
kubelet version: v1.25.3

![image](https://user-images.githubusercontent.com/10864240/202201867-71838a4f-9715-44f7-9708-13858bfb6962.png)

We think it can be related to cgroup v2 but this is all that we have at the moment.

We have the same issue with java and node services running on k3s.

Do you know what can be an issue here? 

Thanks!

**Comment by strainovic** [comment]
When I list files from `/sys/fs/cgroup/`, I noticed that `microk8s` has a few files more:

hugetlb.1GB.current
hugetlb.1GB.events
hugetlb.1GB.events.local
hugetlb.1GB.max
hugetlb.1GB.rsvd.current
hugetlb.1GB.rsvd.max

Maybe this can be some clue?

**Comment by brandond** [comment]
You've asked `stress` to run two workers, each using 250mb for a total of 500mb. I would expect this configuration to consume more than the limit, and get OOM killed. Perhaps a better question would be why you DO NOT see this under microk8s, as opposed to why you DO see this under k3s.

**Comment by strainovic** [comment]
> You've asked `stress` to run two workers, each using 250mb for a total of 500mb. I would expect this configuration to consume more than the limit, and get OOM killed. Perhaps a better question would be why you DO NOT see this under microk8s, as opposed to why you DO see this under k3s.

Hi @brandond,

Hmm, this is really strange, I didn't notice that vm argument is set to 2. In that case, you are right, k3s works as expected.

Besides this test case, we have, as I already mentioned, a few more services written in java and node that behave the same on k3s. For example, Camunda BPMN service uses ~350Mi of memory on microk8s and on other Kubernetes distributions such as AWS EKS and OpenShift, while on k3s, the same service uses ~1Gi of memory.

Not sure how to explain that?

**Comment by brandond** [comment]
I have no idea what your app is doing, and without any access to the image itself it's really hard to guess at why this might be, but I'm not aware of any issues with memory accounting being incorrect under K3s. I would suggest profiling the app under both environments to see if it is actually allocating more or less memory, or if the accounting is just different under whatever container runtime you're using on the other clusters.

**Comment by strainovic** [comment]
Thank you @brandond, I think I've found out the reason, at least for this stress image. Microk8s has swapon by default. After I disabled it with `sudo swapoff -a`, it started to behave the same as k3s.

Back to the lab :)

**Comment by strainovic** [comment]
After additional research, I've found out that an older version of the adoptOpenJDK v11 was the issue. Cgroup v2 support is backported on version 11.0.16, while I had 11.0.14. After rebuilding the image, memory consumption is returned to normal.

Also, I tried to disable cgroup v2 and it also helped to reduce memory consumption, but upgrading was s better solution.

Thank you very much and sorry for blaming k3s :(
